File: cpp/0216组合总和Ⅲ.cpp
#include<vector>

using namespace std;

vector<vector<int>> res;
vector<int> temp;
int sum = 0;

void BackTracking(int s, int k, int n)
{
    if(sum>n)
        return;
    if (k < 1)
    {
        if(sum==n)
            res.push_back(temp);
        return;
    }

    for (int i = s;i<10;++i)
    {
        sum += i;
        temp.push_back(i);
        BackTracking(i + 1, k - 1, n);
        temp.pop_back();
        sum -= i;
    }
}

vector<vector<int>> CombinationSum3(int k,int n)
{
    BackTracking(1, k, n);
    return res;
}

File: python/main/0216组合总和Ⅲ.py


def combinationSum3(k: int, n: int) -> list:
    res=[]
    ss=0
    end=min(10,n)
    
    def _back_tracking(s:int,k:int,n:int,temp:list):
        nonlocal ss
        if(ss>n):
            return
        if(k<1):
            if(ss==n):
                res.append(temp.copy())
            return


        for i in range(s,end-k+1):
            ss+=i
            temp.append(i)
            _back_tracking(i+1,k-1,n,temp)
            temp.pop()
            ss-=i

    _back_tracking(1,k,n,[])

    return res

print(combinationSum3(3,7))